modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Near/NearCache.cs

```diff
@@ -0,0 +1,261 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apache.Ignite.Core.Impl.Cache.Near
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using Apache.Ignite.Core.Cache;
+    using Apache.Ignite.Core.Cache.Affinity;
+    using Apache.Ignite.Core.Impl.Binary;
+    using Apache.Ignite.Core.Impl.Binary.IO;
+
+    /// <summary>
+    /// Holds near cache data for a given cache, serves one or more <see cref="CacheImpl{TK,TV}"/> instances.
+    /// </summary>
+    internal sealed class NearCache<TK, TV> : INearCache
+    {
+        /** Affinity. */
+        private readonly CacheAffinityImpl _affinity;
+        
+        /** Keep binary flag. */
+        private readonly bool _keepBinary;
+
+        /** Topology version func. Returns boxed <see cref="AffinityTopologyVersion"/>.
+         * Boxed copy is passed directly to <see cref="NearCacheEntry{T}"/>, avoiding extra allocations.
+         * This way for every unique <see cref="AffinityTopologyVersion"/> we only have one boxed copy,
+         * and we can update <see cref="NearCacheEntry{T}.Version"/> atomically without locks. */
+        private readonly Func<object> _affinityTopologyVersionFunc;
+
+        /** Underlying map. */
+        private readonly ConcurrentDictionary<TK, NearCacheEntry<TV>> _map = 
+            new ConcurrentDictionary<TK, NearCacheEntry<TV>>();
+
+        /** Stopped flag. */
+        private volatile bool _stopped;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NearCache{TK, TV}"/> class.
+        /// Called via reflection from <see cref="NearCacheManager.CreateNearCache"/>. 
+        /// </summary>
+        public NearCache(Func<object> affinityTopologyVersionFunc, CacheAffinityImpl affinity, bool keepBinary)
+        {
+            _affinityTopologyVersionFunc = affinityTopologyVersionFunc;
+            _affinity = affinity;
+            _keepBinary = keepBinary;
+        }
+
+        /** <inheritdoc /> */
+        public bool IsStopped
+        {
+            get { return _stopped; }
+        }
+
+        public bool TryGetValue<TKey, TVal>(TKey key, out TVal val)
+        {
+            if (_stopped)
+            {
+                val = default(TVal);
+                return false;
+            }
+
+            NearCacheEntry<TV> entry;
+            var key0 = (TK) (object) key;
+            
+            if (_map.TryGetValue(key0, out entry))
+            {
+                if (IsValid(entry))
+                {
+                    val = (TVal) (object) entry.Value;
```

Comment:
You mentioned no-boxing approach, am I missing something?

Reply:
Yes, I mentioned no-boxing approach regarding `_map` when proper types are specified in `PlatformNearCacheConfiguration`.

This line is just a cast from `TV` to `TVal`, which can't be done any other way. Previously we used `TypeCaster` class to avoid boxing in such cases (see usages), and it was proved to be better on .NET 4. However, this is no longer the case on .NET Core:

```
Runtime=.NET Core 3.1  

|     Method |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |----------:|----------:|----------:|------:|------:|------:|----------:|
| TypeCaster | 1.5402 ns | 0.0138 ns | 0.0122 ns |     - |     - |     - |         - |
| ObjectCast | 0.0284 ns | 0.0098 ns | 0.0087 ns |     - |     - |     - |         - |
```

(source: https://github.com/ptupitsyn/IgniteNetBenchmarks/blob/master/CasterBenchmark.cs)

As you can see, such a cast is optimized by JIT, there are zero allocations, and it is faster than `TypeCaster`.

(There is a ticket to get rid of TypeCaster: https://issues.apache.org/jira/browse/IGNITE-12699)